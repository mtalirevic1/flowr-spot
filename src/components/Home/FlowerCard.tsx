import {FC, useState} from 'react'
import {Box, IconButton, Typography, useMediaQuery, useTheme} from '@mui/material'
import {Flower} from '../../model/flower'
import StarIcon from '@mui/icons-material/Star'

type FlowerCardProps = {
    flower: Flower
    isLoggedIn: boolean
}

const FlowerCard: FC<FlowerCardProps> = ({flower, isLoggedIn}) => {
    const {name, latin_name, profile_picture, favorite, sightings} = flower

    const theme = useTheme()
    const isMobile = useMediaQuery(theme.breakpoints.only('xs'))

    const [isFavorite, setIsFavorite] = useState(favorite)

    return (
        <div>
            <Box sx={{
                backgroundImage: `url(${profile_picture})`,
                backgroundRepeat: 'no-repeat',
                backgroundSize: 'cover',
                backgroundPosition: 'center',
                width: {
                    xs: '184px',
                    sm: '200px',
                    md: '232px',
                    lg: '280px',
                    xl: '280px'
                },
                height: {
                    xs: '230px',
                    sm: '250px',
                    md: '290px',
                    lg: '350px',
                    xl: '350px'
                },
                borderRadius: '3px',
            }}>
                <Box
                    sx={{
                        width: '100%',
                        height: '100%',
                        background: 'linear-gradient(180deg, rgba(0, 0, 0, 0.0001) 0%, rgba(0, 0, 0, 0.7) 89.5%)',
                        backgroundRepeat: 'no-repeat',
                        backgroundSize: 'cover',
                        backgroundPosition: 'center',
                        display: 'flex',
                        justifyContent: 'flex-end',
                        alignItems: 'center',
                        flexDirection: 'column',
                        borderRadius: '3px',
                    }}
                >
                    {isLoggedIn &&
                        <IconButton
                            size={isMobile ? 'small' : 'medium'}
                            onClick={() => setIsFavorite(prevFav => !prevFav)}
                            sx={{
                                position: 'relative',
                                top: {
                                    xs: '-40%',
                                    sm: '-33%',
                                    md: '-40%',
                                    lg: '-50%',
                                    xl: '-55%'
                                },
                                right: {
                                    xs: '-35%',
                                    sm: '-33%',
                                    md: '-33%',
                                    lg: '-40%',
                                    xl: '-39%'
                                },
                                background: (theme) => isFavorite ? theme.palette.primary.main : 'white',
                                '&:hover': {
                                    background: (theme) => isFavorite ? theme.palette.primary.main : 'white',
                                }
                            }}>
                            <StarIcon
                                fontSize={isMobile ? 'small' : 'medium'}
                                sx={{
                                    color: (theme) => isFavorite ? 'white' : theme.palette.neutral.dark,
                                    '&:hover': {
                                        color: (theme) => isFavorite ? 'white' : theme.palette.neutral.dark,
                                    }
                                }}
                            />
                        </IconButton>
                    }
                    <Typography
                        variant={isMobile ? 'h4' : 'h3'}
                        fontWeight={400}
                        color="white"
                    >
                        {name}
                    </Typography>
                    <Typography
                        fontSize={isMobile ? '12px' : '9.6px'}
                        fontWeight={400}
                        fontStyle="italic"
                        color="white"
                        sx={{opacity: 0.7, mt: '3px'}}
                    >
                        {latin_name}
                    </Typography>
                    <Typography
                        mb={isMobile ? '15px' : '20px'}
                        mt={isMobile ? '15px' : '20px'}
                        p={isMobile ? '6px' : '9px 15px'}
                        borderRadius={isMobile ? '16px' : '20px'}
                        color="white"
                        sx={{
                            background: (theme) => isFavorite && isLoggedIn ?
                                theme.palette.primary.main : 'rgba(0, 0, 0, 0.7)'
                        }}
                    >
                        {`${sightings} sightings`}
                    </Typography>
                </Box>
            </Box>
        </div>
    )
}

export default FlowerCard