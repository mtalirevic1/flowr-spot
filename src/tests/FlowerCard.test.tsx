import {it, describe} from 'vitest'
import {render, screen} from '@testing-library/react'
import renderer from 'react-test-renderer'

import FlowerCard from '../components/Home/FlowerCard'

const testFlower = {
    id: 1,
    name: 'Test',
    latin_name: 'Testus Testati',
    profile_picture: '',
    sightings: 4,
    favorite: false
}

describe('FlowerCard', () => {
    it("should show the favorite button while user is logged in", () => {
        render(<FlowerCard
            flower={testFlower}
            isLoggedIn={true}/>)

        const favButton = screen.getByTestId('favoriteButton')

        expect(favButton).toBeVisible()
    })
})

describe('FlowerCard Snapshot', ()=>{
    it('should match DOM Snapshot', () => {
        const tree = renderer.create(<FlowerCard flower={testFlower} isLoggedIn={true}/>).toJSON()
        expect(tree).toMatchSnapshot()
    })
})