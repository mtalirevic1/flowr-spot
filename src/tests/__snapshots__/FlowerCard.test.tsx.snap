// Vitest Snapshot v1

exports[`FlowerCard Snapshot > should match DOM Snapshot 1`] = `
<div>
  <div
    className="MuiBox-root css-cqcczv"
  >
    <div
      className="MuiBox-root css-1cfdb0z"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-qxah7x-MuiButtonBase-root-MuiIconButton-root"
        data-testid="favoriteButton"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-lw0a3y-MuiSvgIcon-root"
          data-testid="StarIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
          />
        </svg>
      </button>
      <h3
        className="MuiTypography-root MuiTypography-h3 css-1r4bjj5-MuiTypography-root"
      >
        Test
      </h3>
      <p
        className="MuiTypography-root MuiTypography-body1 css-9gn7k5-MuiTypography-root"
      >
        Testus Testati
      </p>
      <p
        className="MuiTypography-root MuiTypography-body1 css-17hg2cb-MuiTypography-root"
      >
        4 sightings
      </p>
    </div>
  </div>
</div>
`;
